core/src/main/java/org/apache/calcite/rel/rel2sql/RelToSqlConverter.java

```diff
@@ -504,12 +518,16 @@ protected Builder buildAggregate(Aggregate e, Builder builder,
           SqlStdOperatorTable.ROLLUP.createCall(SqlParserPos.ZERO, groupKeys));
     default:
     case OTHER:
+      List<SqlNode> groupingSetsList = aggregate.getGroupSets().stream()
+          .map(groupSet -> groupItem(groupKeys, groupSet, aggregate.getGroupSet()))
+          .collect(Collectors.toList());
+      if (ImmutableBitSet.union(aggregate.getGroupSets()).cardinality()
+          != aggregate.getGroupSet().cardinality()) {
+        groupingSetsList.add(
```

Comment:
According to _Aggregate_ constructor doc, the grouping sets must be strictly ordered by inclusion, so the superset grouping set should always be the first.

Reply:
Good point. Strictly, that only applies to the `Aggregate.groupSets` data structure (list of `ImmutableBitSet`), not the `SqlNodeList`, but it's a good idea to follow the same pattern. I will make the change.

Reply:
done